lib/inspec/shell_detector.rb

```diff
@@ -0,0 +1,90 @@
+# encoding: utf-8
+require 'etc'
+require 'rbconfig'
+
+module Inspec
+  #
+  # ShellDetector attempts to detect the shell the invoking user is
+  # running by checking:
+  #
+  #  - The command of our parent
+  #  - The SHELL environment variable
+  #  - The shell returned by getpwuid for our process UID
+  #
+  # Since none of these methods is fullproof, the detected shell is
+  # verified against a list of known shells before being returned to
+  # the caller.
+  #
+  class ShellDetector
+    NOT_DETECTED = Object.new.freeze
+    KNOWN_SHELLS = %w{bash zsh ksh csh sh fish}.freeze
+
+    def initialize
+      @shell = NOT_DETECTED
+    end
+
+    def shell
+      @shell = detect if !detected?(@shell)
+      @shell
+    end
+
+    def shell!
+      @shell = detect
+    end
+
+    private
+
+    def detect
+      # Most of our detection code assumes a unix-like environment
+      return nil if (RbConfig::CONFIG['host_os'] =~ /mswin|mingw|cygwin/)
```

Comment:
May it make sense to output a warning? Otherwise the user may be confused that nothing has been executed?